File: ts/src/adapters/kimi-code.ts
import { register } from '../registry.js'
import { HarnessError, finalizeCommand, validateRunSpec } from '../base.js'
import type { Adapter, BuildCommand, ParsedOutput, RunSpec, SubprocOutcome } from '../base.js'
import { parseJsonObjectLines } from '../util.js'

// Kimi Code CLI 0.41.0 (`@moonshot-ai/kimi-code`, release commit 95478e8c)
// runs headless once `--prompt` is given and, with `--output-format
// stream-json`, writes one JSON object per stdout line: assistant messages
// (content/tool_calls), tool results (tool_call_id/content) and meta records.
// Print mode is inherently auto-approving, so there is no bypass mapping and
// the common validator rejects `permissionPolicy: 'bypass'`. No accounting
// schema is qualified, so metrics stay null and objects are preserved verbatim.

const kimiCodeAdapter: Adapter = {
  name: 'kimi-code',
  instructionsFilename: 'AGENTS.md',
  defaultModel: '', // Leave model selection to the caller's Kimi config.
  configHomeEnv: 'KIMI_CODE_HOME',

  buildCommand(spec: RunSpec): BuildCommand {
    const validated = validateRunSpec(this, spec)
    // Upstream rejects a blank prompt (`prompt.trim().length === 0`); without
    // `--prompt` the CLI opens its interactive session instead.
    if (!spec.prompt?.trim()) {
      throw new HarnessError('kimi-code requires a non-blank prompt for print mode', 'invalid-options')
    }
    const { model } = validated
    const args = ['--output-format', 'stream-json']
    // Equals form keeps values starting with '-' from being parsed as flags.
    if (model) args.push(`--model=${model}`)
    args.push(`--prompt=${spec.prompt}`)
    return finalizeCommand(this, spec, validated, { cmd: 'kimi', args, model: spec.model || null })
  },

  parseOutput(_spec: RunSpec, outcome: SubprocOutcome): ParsedOutput {
    const events = parseJsonObjectLines(outcome.stdout)
    return { costUsd: null, tokensIn: null, tokensOut: null, raw: events.length === 0 ? null : events }
  },
}

kimiCodeAdapter.installMeta = {
  packageManager: 'npm',
  installCommand: ['npm', 'install', '-g', '@moonshot-ai/kimi-code'],
  updateCommand: ['npm', 'install', '-g', '@moonshot-ai/kimi-code@latest'],
  versionCommand: ['kimi', '--version'],
}

register('kimi-code', kimiCodeAdapter)

File: src/harness/adapters/kimi_code.py
"""Kimi Code CLI (`kimi`) 0.41.0 print mode; see ADAPTER-MATRIX.md.

`kimi --output-format stream-json --prompt=<text>` runs one non-interactive
turn and emits assistant / tool / meta JSON objects as JSON Lines. No usage or
billing schema is qualified, so every cost and token metric stays None and each
complete JSON object line is preserved in `raw` (None when the stream carried
none).

The model is selected with `--model=<alias>` only when requested; the default
leaves selection to the caller's Kimi config, and the exact alias is passed
through untouched. Print mode is inherently auto-approving and rejects the
interactive `--yolo` / `--auto` / `--plan` mode flags, so only the upstream
permission policy exists. `KIMI_CODE_HOME` relocates the config home; there is
no config-file flag.
"""
from __future__ import annotations

from harness._subproc import SubprocOutcome
from harness.base import Adapter, BuildCommand, HarnessError, InstallMeta, ParsedOutput, RunSpec
from harness.util import json_object_events


class KimiCodeAdapter(Adapter):
    name = "kimi-code"
    instructions_filename = "AGENTS.md"
    DEFAULT_MODEL = ""  # Leave model selection to the caller's Kimi config.
    config_home_env = "KIMI_CODE_HOME"
    install_meta = InstallMeta(
        package_manager="npm",
        install_command=("npm", "install", "-g", "@moonshot-ai/kimi-code"),
        update_command=("npm", "install", "-g", "@moonshot-ai/kimi-code@latest"),
        version_command=("kimi", "--version"),
    )

    def reported_model(self, spec: RunSpec) -> str | None:
        return spec.model or None

    def build_command(self, spec: RunSpec) -> BuildCommand:
        resolved = self.resolve_run_spec(spec)
        if not spec.prompt.strip():
            # Upstream rejects a blank --prompt and would otherwise open the interactive UI.
            raise HarnessError("kimi-code requires a non-empty prompt for print mode", code="invalid-options")
        args = ["--output-format", "stream-json"]
        if resolved.model:
            # Equals form keeps leading '-' model names from being parsed as flags.
            args.append(f"--model={resolved.model}")
        args.append(f"--prompt={spec.prompt}")
        return self.finalize_command(spec, cmd="kimi", args=args)

    def parse_output(self, spec: RunSpec, outcome: SubprocOutcome) -> ParsedOutput:
        return {"cost_usd": None, "tokens_in": None, "tokens_out": None, "raw": json_object_events(outcome.stdout)}
